rfcs/20240326-block-level-api/README.md

```diff
@@ -0,0 +1,540 @@
+- support runtime leading dimensions.
+
+The proposal here is to stick with what is already tried and used
+internally. If during the experimental stage of the API, users request
+these functionalities, we will revisit the API and add the necessary
+support in the internal implementation. So alpha/beta will be creation
+time parameters, and kernels with and without post-ops will be handled
+with separate objects.  For runtime leading dimension, the internal
+API supports it but this will not be exposed for now. We can either
+change the API when we introduce the feature, or add a placeholder
+parameter if we are confident this feature will gain traction soon.
+
+Also, note that currently, no kernel-level cache exists for block
+level functionalities, but there are plans to add it. In the meantime,
+here are a few ways to mitigate the lack of kernel-level cache:
+- hoist block-level kernel generation for it to happen once at the
+  beggining of the application
+- to handle tails/K-blocking, if generating multiple kernels is too
+  costly, one can copy A/B blocks to a temporary buffer for
+  tails/first K-block, and rely on a single block-level kernel.
+
+
+## Transforms and transpose
+
+Transormation rountines, for example to pack data in a certain layout,
+are typically hard to implement without using intrinsics or assembly.
+To facilitate packing, we will expose an out-of-place transform
+functionality.
+
+## All-in-all
+
+### Interface proposal
+
+```c++
+
+// namespace name to be defined, leaving it general enough for additional block level APIs
+namespace dnnl {
+namespace block { 
+
+enum packing_tag {
+packed_32;
+packed_64;
+plain;
+transposed;
+}
+
+struct brgemm {
+    struct desc {
+        // Vanilla version of brgemm with no post-op or destination conversion.
+        desc(dim_t batch, dim_t M, dim_t N, dim_t K,
+                data_type dtA, dim_t ldA,
+                data_type dtB, dim_t ldB, 
+                data_type dtC, dim_t ldC,
+                float alpha, float beta);
+
+        // Advanced version with postop and datatype conversion when D type
+        // is different than C type.
+        desc(dim_t batch, dim_t M, dim_t N, dim_t K,
+                data_type dtA, dim_t ldA,
+                data_type dtB, dim_t ldB, 
+                data_type dtC, dim_t ldC,
+                data_type dtD, dim_t ldD,
+                float alpha, float beta, 
+                const brgemm_attr &attr);
+
+        // Queries for expected layouts and temporary memory
+        packing_tag get_A_tag() const;
+        packing_tag get_B_tag() const;
+        size_t get_scratchpad_size() const;
+    }
+
+    brgemm(const desc &bd); 
+    
+    // HW context handling.
+    // This currently mimics AMX (need to clarify for SME): 
+    // - Release is static
+    // - No release necessary between different calls to set_hw_context
+    void set_hw_context() const;
+    static void release_hw_context() const;
```

Comment:
1. "set_hw_context()" seems very generic (maybe intentionally). But it raises a question about what is being modified (mxcsr, etc.. ). Maybe provide a documentation of what features it is allowed to modify, for transparency?
2. By static do you mean, it is safe to call it regardless of a previous call to `set_hw_context()`?

Reply:
1. this is a valid point and I left it fuzzy on purpose :-). In general it just changes the state of the core the function is executed on. If need be, we can be more specific in documentation and specify for each ISA, what is being changed.
2. just that it can be called without a brgemm object. This is to highlight that releasing HW context should be independent of how it was set.